experimental/formatted_string/src/lib.rs

```diff
@@ -6,9 +6,128 @@
 
 extern crate alloc;
 
-mod error;
 mod formatted_string;
 
 pub use crate::formatted_string::FormattedString;
-pub use crate::formatted_string::FormattedStringLike;
-pub use error::Error as FormattedStringError;
+
+use core::fmt;
+pub use writeable::{LengthHint, Writeable};
+
+/// Similar to Writeable, but also produces field annotations
+pub trait FormattedWriteable {
+    /// Write bytes and field annotations to the given FormattedWriteableSink. Errors from
+    /// the sink are bubbled up.
+    fn fmt_write_to<S: FormattedWriteableSink>(&self, sink: &mut S) -> Result<(), S::Error>;
+
+    /// Returns a hint for the number of bytes that will be written to the sink.
+    ///
+    /// Override this method if it can be computed quickly.
+    fn write_len(&self) -> LengthHint {
+        LengthHint::undefined()
+    }
+
+    fn writeable_to_fmt_string(&self) -> FormattedString {
+        let mut output = FormattedString::with_capacity(self.write_len().capacity());
+        self.fmt_write_to(&mut output)
+            .expect("impl FormattedSink for FormattedString is infallible");
+        output
+    }
+
+    fn as_writeable(self) -> FormattedWriteableAsWriteable<Self>
+    where
+        Self: Sized,
+    {
+        FormattedWriteableAsWriteable(self)
+    }
+}
+
+/// A sink for a FormattedWriteable. See FormattedString for an implementation.
+pub trait FormattedWriteableSink {
+    type Error: core::fmt::Display;
```

Comment:
If we had a concrete error type, what would it be? 

In this case I feel like an associated error type is fine but unsure

Reply:
It would need to be `core::fmt::Error` for `CoreWriteAsFormattedWriteableSink` to work, which means we couldn't use `Infallible` for `FormattedString`.